### DISCUSSION TITLE
Collection expressions: Allow builder-pattern method to return different type than the type being built.

### DISCUSSION THREAD

**Discussion by CyrusNajmabadi** [dotnet/csharplang#7398] [Language Ideas]
## Proposal

This issue contains a change request to the current [collection expressions](https://github.com/dotnet/csharplang/blob/main/proposals/collection-expressions.md) specification that the working group would like to bring for LDM consideration.

The core use case we want to support is relates to the following APIs in the BCL.  First, these types, with these members exist today:

```c#
public interface IImutableList<T> { /* ... */ }

[CollectionBuilder(typeof(ImmutableList), "Create")]
public class ImmutableList<T> : IImutableList<T> { /* ... */  }

public static class ImmutableList
{
    public static ImmutableList<T> Create<T>(ReadOnlySpan<T> values);
}
```

With the above, the user *can* write the following:

```c#
public class C
{
    public ImutableList<X> Values { get; } = [GetValue(a), GetValue(b), GetValue(c)];
}
```

But they *cannot* write:

```c#
public class C
{
    // this is the interface, not the concrete type
    public IImutableList<X> Values { get; } = [GetValue(a), GetValue(b), GetValue(c)];
}
```

This doesn't work because `IImutableList<T>` is currently not something the compiler knows how to create.  This is unfortunate given such a canonical implementation.  Fortunately, a solution seems readily available.  Specifically, update the interface like so:

```diff
+ [CollectionBuilder(typeof(ImmutableList), "Create")]
public interface IImutableList<T> { /* ... */ }
```

Bringing it in line with the class.  With this, the compiler would now be able to find `ImmutableList.Create` as the way to initialize `C.Values`.   *However*, as the spec is currently written, this would not work. Specifically, the [specification says](https://github.com/dotnet/csharplang/blob/main/proposals/collection-expressions.md#create-methods): 

> For the create method:
> - ...
> - There is an [identity conversion](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/conversions.md#1022-identity-conversion) between the method return type and the collection type.

Here the method return type is `ImmutableList<T>` (the concrete type) while the collection type is `IImutableList<T>` (the interface type).   This is not an identity conversion, so this is not allowed.

However, given how trivial this is to support, and given that the BCL is fine adding these attributes to these collections ([see PR here](https://github.com/dotnet/runtime/pull/89459)), we propose tweaking the above rule to one of either:

1. There is an *implicit conversion* between the method return type and the collection type.

In other words, allow the user's code above because it would be equivalent to:

```c#
public class C
{
    public IImutableList<X> Values { get; } = ImmutableList.Create(GetValue(a), GetValue(b), GetValue(c));
}
```

Given that the full form would be legal to write there, requiring an exact implicit conversion, instead of any implicit conversion is just onerous.

Costs here are minimal.  BCL PR is already done, and runtime has indicated they are fine with this sort of pattern. Compiler PR would just be a minor tweak to the check, along with tests to validate behavior.  

## Alternatives

One suggestion would be to add the `[CollectionBuilder(typeof(...), "Create")]` attribute to tehse interfaces, and have them point at methods that *do* return an instance of the exact interface type the attribute is on.  

The runtime team is against this.  They would be creating additional, highly redundant, surface area just to work around this limitation in the current specification.  The methods would be clones (or dispatchers) to the existing methods, and would need to find some new place to live, or would live next to the existing Create methods, just with different names.  This would be extra code, apis, and docs for methods users themselves would never be expected to call.

Conversely, the runtime *already* has the existing `Create` methods which work perfectly for our needs, if we're just willing to call them and not have this restriction interfere.  So that seems to be the simplest, clearest, and most desirable path for both teams.

**Comment by bernd5** [comment]
That would be great.
But ff we don't want to change the current proposal, the runtime team could specify another Create method e.g. `CreateInterface`.

So we would get:

```csharp
public static class ImmutableList
{
    public static IImmutableList<T> CreateInterface<T>(ReadOnlySpan<T> values) => Create<T>(values);
    public static ImmutableList<T> Create<T>(ReadOnlySpan<T> values){  
}

[CollectionBuilder(typeof(ImmutableList), "CreateInterface")]
public interface IImutableList<T> { /* ... */ }

[CollectionBuilder(typeof(ImmutableList), "Create")]
public class ImmutableList<T> : IImutableList<T> { /* ... */  }
```

**Reply by CyrusNajmabadi** [reply]
> But ff we don't want to change the current proposal, the runtime team could specify another Create method e.g. CreateInterface.

The runtime team expressed *explicitly* that they did not want to do that.  :)

They already have the method.  It already has the appropriate shape on inputs/outputs.  Adding another method *only* to workaround the lack of the language not being willing to use the existing method is not something there is willingness to do :)

**Reply by jnm2** [reply]
I think it would be great to mention the runtime's position in the change request itself, too.

**Reply by CyrusNajmabadi** [reply]
will add a section.

**Comment by KennethHoff** [comment]
I don't see a reason why not to allow this.

Btw, is there a requirement that `Collection Expressions` have to create `IEnumerable`s, or could I use it for any arbitrary type, as long as we add the method to create it from a ReadonlySpan?

**Reply by CyrusNajmabadi** [reply]
> Btw, is there a requirement that Collection Expressions have to create IEnumerables, 

There is no such requirement.  You can see the specification here https://github.com/dotnet/csharplang/blob/main/proposals/collection-expressions.md for how it works.

High level intuition:

1. if your type already supports the collection-initializer pattern (e.g. `new X { 1, 2, 3 }`) it will work with collection expressions.
2. If your type does not support that pattern, you can use an attribute to point at a suitable `Create(ReadOnlySpan<T>)` method to have hte compiler construct instances on your behalf.  

This discussion is about a slight relaxing of '2' to support a broader set of cases, including ones that arise in the BCL itself.

**Reply by jnm2** [reply]
The requirement is that the type being created by the collection expression is foreachable so that we know what the element type of the collection expression is. You could technically be foreachable without being IEnumerable, but you still have to implement the shape of IEnumerable if not the interface itself.

That means you're going to be happy if your type is enumerable in nature but can't implement interfaces (ref struct). And sad if your type is not supposed to be enumerable in nature, because any type that can't be enumerated with `foreach` can't be created via collection expression.

**Reply by KennethHoff** [reply]
>has to be foreachable

That seems like a weird constraint.

This is just a random example I came up with, but imagine I wanted to create a "DateTime ranges expression" using this syntax:

```c#
private static readonly CustomDateRange dateRange = [(2022..2024), (03..08), (12..18)];
```

This `CustomDateRange` doesn't make sense to be "looped through". This might just be a bastardization of the syntax though, and better served by just the constructor, as it would break if you didn't put in exactly 3 values.

```c#
private static readonly CustomDateRange dateRange = new(2022..2024, 03..08, 12..18);
```

The more I think about it, the worse this example is, but I wrote this on my phone, so I'll keep it.. 😭

**Reply by jnm2** [reply]
I replied while forgetting that foreachability is not currently required when `[CollectionBuilder]` is not applied. I do feel strange that an optimization (switching to a Create method instead of Add calls) results in a semantics change (requiring foreachability instead of not requiring that).

My understanding is that the reason we specced [CollectionBuilder] to require foreachability was to simplify the design, along with other simplifications that whittle down unlikely scenarios. We could relax the rule later if there was sufficient demand.

**Reply by huoyaoyuan** [reply]
Note that spans are not IEnumerable since they can't implement interface now, but they are foreachable. Memory is not IEnumerable either.

**Reply by CyrusNajmabadi** [reply]
@huoyaoyuan thanks.  'Span' doesn't matter here as it's implicitly creatable by collection literals.  'Memory' is not a collection type (and there's no movement to make it one) so it doesn't apply either.  Thanks!

**Reply by jnm2** [reply]
@CyrusNajmabadi You could think of Memory as "async-capable Span." If the target type to some API is Memory, we could pass an array.

**Reply by jnm2** [reply]
This has become a working group discussion. I really appreciate the reminder about System.Memory. A type in a similar boat (implicit conversion from array) is System.ArraySegment.

**Reply by gulbanana** [reply]
I have a use case for [CollectionBuilder] without foreachability. I've got an ORM type which looks something like this:
```cs
interface IUnloadedSet<T>
{
    void Add(T);
    void Remove(T);
    Task<ILoadedSet<T>> LoadAsync();
}

interface ILoadedSet<T> : IEnumerable<T>
{
    bool Contains(T);
}
```

Because the non-Loaded interface isn't IEnumerable, we can't create it with collection initializers, even though it has Add(). However, the desired semantics are that the sets aren't *necessarily* loaded - initializing them with data that happens to be loaded is fine. It would be great to be able to put [CollectionBuilder] on it.

**Reply by jnm2** [reply]
Thanks for the example! Let's continue to see if this is a common enough need. My personal leaning currently is that it feels a little like a concept mismatch, but I get the convenience part of it and maybe seeing examples would change my leaning.

In the meantime, the workaround would be to pass the collection expression to a helper method which returns your interface type.

**Reply by gulbanana** [reply]
Pre-collection expressions, my workaround is collection initializers on a concrete instance of the loaded type. It looks like this:
```cs
new ORMEntity
{
    Collection = new LoadedSet<OtherEntity>() { otherEntity1, otherEntity2 }
}
```

I imagine the helper method version could be something like
```cs
new ORMEntity
{
    Collection = Set.Create([otherEntity1, otherEntity2])
}
```

**Reply by jnm2** [reply]
@gulbanana Can you share your use case in <https://github.com/dotnet/csharplang/issues/7744>?

**Reply by jnm2** [reply]
> 'Memory' is not a collection type (and there's no movement to make it one)

This has now changed: https://github.com/dotnet/csharplang/issues/8144
